lib/msal-common/test/client/ClientCredentialClient.spec.ts

```diff
@@ -359,5 +360,61 @@ describe("ClientCredentialClient unit tests", () => {
 
         await expect(client.acquireToken(clientCredentialRequest)).rejects.toMatchObject(ClientAuthError.createMultipleMatchingTokensInCacheError());
     });
+    it("Uses the extensibility AppTokenProvider callback to get a token", async () => {
+        sinon.stub(Authority.prototype, <any>"getEndpointMetadataFromNetwork").resolves(DEFAULT_OPENID_CONFIG_RESPONSE.body);
+        // no need to stub out the token response, MSAL will use the AppTokenProvider instead
+
+        const config = await ClientTestUtils.createTestClientConfiguration();
+        const accessToken = "some_token";
+        const appTokenProviderResult: AppTokenProviderResult = {
+            accessToken: accessToken,
+            expiresInSeconds: 1800,
+            refreshInSeconds: 900,
+        }
+
+        const expectedScopes = [TEST_CONFIG.DEFAULT_GRAPH_SCOPE[0]];
+
+        let callbackedCalledCount = 0;
+
+        config.appTokenProvider =  (appTokenProviderParameters) => {
+            callbackedCalledCount++;
+
+            expect(appTokenProviderParameters.scopes).toEqual(expectedScopes);
+            expect(appTokenProviderParameters.tenantId).toEqual("common");
+            expect(appTokenProviderParameters.correlationId).toEqual(TEST_CONFIG.CORRELATION_ID);
+            expect(appTokenProviderParameters.claims).toBeUndefined();
+                
+            return new Promise<AppTokenProviderResult>(
+                (resolve) => resolve(appTokenProviderResult));                        
+        };
+    
+        // client credentials not needed
+        config.clientCredentials = undefined;    
+
+        const client = new ClientCredentialClient(config);
+        const clientCredentialRequest: CommonClientCredentialRequest = {
+            authority: TEST_CONFIG.validAuthority,
+            correlationId: TEST_CONFIG.CORRELATION_ID,
+            scopes: TEST_CONFIG.DEFAULT_GRAPH_SCOPE,
+        };
+
+        const authResult = await client.acquireToken(clientCredentialRequest) as AuthenticationResult;
+
+        expect(callbackedCalledCount).toEqual(1);
 
+        expect(authResult.scopes).toEqual(expectedScopes);
+        expect(authResult.accessToken).toEqual(accessToken);
+        expect(authResult.state).toHaveLength(0);
+        const dateDiff = (authResult.expiresOn!.valueOf() - Date.now().valueOf()) / 1000;
+        expect(dateDiff).toBeLessThanOrEqual(1900);
+        expect(dateDiff).toBeGreaterThan(1700);
```

Comment:
So the test fails if it takes more than 200ms to run?

Reply:
Those numbers are in seconds, not ms. I wanted a way to test that the expiration is the one set by the test at line 371, i.e. 1800s. I can't use exactly 1800 in the assert because some time will have elapsed. So, I put a generous interval in there.